pkg/controller/iam/serviceaccountkey.go

```diff
@@ -0,0 +1,336 @@
+
+	if err != nil {
+		return errors.Wrap(err, errDeleteServiceAccountKey)
+	}
+
+	if !extNameExists {
+		return nil
+	}
+
+	_, err = s.serviceAccountKeyClient.Delete(rrn).Context(ctx).Do()
+
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, errDeleteServiceAccountKey)
+}
+
+// NewRelativeResourceNamer makes an instance of the RelativeResourceNamer
+// which is the only type that is allowed to know how to construct GCP resource names
+// for the IAM type.
+func newSaKeyRelativeResourceNamer(projectID string) saKeyrelativeResourceNamer {
+	return saKeyrelativeResourceNamer{
+		saNamer: RelativeResourceNamer{
+			projectName: projectID,
+		},
+	}
+}
+
+// saKeyrelativeResourceNamer allows the controller to generate the Google Cloud API relative resource name
+// for the service account key based on the external-name annotation.
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+// The relative resource names for service account keys have the following format:
+// projects/{project_id}/serviceAccounts/{account}/keys/{key}
+type saKeyrelativeResourceNamer struct {
+	saNamer RelativeResourceNamer
+}
+
+// referencedServiceAccountPath returns the external name of the service account which saKey belongs to
+func (rrn saKeyrelativeResourceNamer) referencedServiceAccountPath(saKey *v1alpha1.ServiceAccountKey) (string, error) {
+	// we assume Google Cloud API relative resource name (aka resource path) of the ServiceAccount is stored in spec.ForProvider.ServiceAccount
+	if saKey.Spec.ForProvider.ServiceAccount == nil || *saKey.Spec.ForProvider.ServiceAccount == "" {
+		return "", fmt.Errorf(fmtErrInvalidServiceAccountRef, saKey.Spec.ForProvider.ServiceAccountReferer)
+	}
+
+	return *saKey.Spec.ForProvider.ServiceAccount, nil
+}
+
+// resourcePath yields the Google Cloud API relative resource name for the ServiceAccountKey resource
+//   returns <the relative resource name>, <whether the external name annotation is non-empty>, <error encountered during resolution>
+func (rrn saKeyrelativeResourceNamer) resourcePath(saKey *v1alpha1.ServiceAccountKey) (string, bool, error) {
+	if saPath, err := rrn.referencedServiceAccountPath(saKey); err != nil {
+		return "", false, err
+	} else if extName := meta.GetExternalName(saKey); extName == "" {
+		return "", false, nil
+	} else {
+		return fmt.Sprintf(fmtKeyRelativeResourceName, saPath, extName), true, nil
+	}
+}
+
+// parseKeyIDFromRrn tries to parse key id from Google Cloud API relative resource name (resource path) of
+//   a service account key
+func parseKeyIDFromRrn(rrn string) (string, error) {
+	resourcePath, err := url.Parse(rrn)
+
+	if err != nil {
+		return "", err
+	}
+
+	return path.Base(resourcePath.Path), nil
+}
+
+func populateSaKey(cr *v1alpha1.ServiceAccountKey, fromProvider *iamv1.ServiceAccountKey) error {
+	keyID, err := parseKeyIDFromRrn(fromProvider.Name)
+
+	if err != nil {
+		return err
+	}
+
+	cr.Status.AtProvider.KeyID = keyID
```

Comment:
I thought it's nice to make this id available as part of resource state because it's a Google API generated id and it's available in the output of "gcloud iam service-accounts keys list".